s2/loop.go

```diff
@@ -1287,11 +1289,26 @@ func (l *Loop) decode(d *decoder) {
 		return
 	}
 	l.vertices = make([]Point, nvertices)
-	for i := range l.vertices {
-		l.vertices[i].X = d.readFloat64()
-		l.vertices[i].Y = d.readFloat64()
-		l.vertices[i].Z = d.readFloat64()
+
+	pointsNeeded := int(nvertices) * 3
+
+	i := 0
+
+	for pointsNeeded > 0 {
+		arr, pointsRead := d.readFloat64Array(pointsNeeded, l.polygon.BufPool)
+		if pointsRead == 0 {
+			break
+		}
+		pointsNeeded = pointsNeeded - pointsRead
+		for j := 0; j < int(pointsRead/3); j++ {
```

Comment:
We have to make sure that pointsRead is a multiple of 3, 
( I am assuming that since you are creating buffers with sizes as multiple of 3, formula: 8*x*3 , so won't run into a situation where pointsRead isn't multiple of 3 )